Draft of SDLC

mbryzek opened the pull request.

mikeclimbrock commented on Sdlc.md:
```diff
@@ -0,0 +1,108 @@
+# Software Development Lifecycle
+
+
+Flow's Software Development Lifecycle (SDLC) is focused on creating an amazing environment for individuals and teams to contribute safely to the growth of Flow. One of the pillars of our culture is trust and empowerment - and our SDLC is informed by and designed to demonstate and further this value.
```
"the growth of Flow" => "Flow's growth"

mikeclimbrock replied:
"One of the pillars" => "One pillar"

mbryzek replied:
added

mikeclimbrock commented on Sdlc.md:
```diff
@@ -0,0 +1,108 @@
+
+## Philosophy
+
+We pursue several practices to ensure our SDLC achieves its goals of providing very high productivity for each of us while continually reducing risk.
```
remove "for each of us"

mbryzek replied:
done

mikeclimbrock commented on Sdlc.md:
```diff
@@ -0,0 +1,108 @@
+
+We pursue several practices to ensure our SDLC achieves its goals of providing very high productivity for each of us while continually reducing risk.
+
+  - We remind ourselves constantly that **failure is a mandatory component of success**. As people, learning and failure simply go hand in hand.
```
remove "simply"

mbryzek replied:
done

mikeclimbrock commented on Sdlc.md:
```diff
@@ -0,0 +1,108 @@
+  
+  - Not all failures are created equal. We use the word **Intentionality** to describe an intentional change to a system. In particular, failures that result from intentional changes are the changes we expect to learn from; by contrast, a lack of intentionality leads to negligence or carelessness which erodes trust and therefore is contrary to our culture.
+  
+  - We **learn and adapt from prior learnings**; In particular, from Kaizen we adopt the learnings of continuous improvement. When failure occurs, we immediately determine what needs to be done to prevent recurrence and focus on those tasks. We learn about antifragile - the notion that following each learning, we emerge stronger as an organization for the future. We learn from Scrum and Cambam wa ...
```
add a "," after Kaizen

mikeclimbrock replied:
"Cambam" => "Kanban"

mbryzek replied:
done

mikeclimbrock commented on Sdlc.md:
```diff
@@ -0,0 +1,108 @@
+  
+  - We focus on software and architecture that continually reduces the risk of changes. These practices focus on:
+  
+    - isolation - truly ensuring that one change cannot impact another. For example, our focus on API design and micro service architecture on the backend is one technique we use to create isolation
```
some of these bullets have periods, and others do not.  make consistent

chrhicks replied:
add periods.. you're actually writing full sentences.

mbryzek replied:
done - all have periods

mikeclimbrock commented on Sdlc.md:
```diff
@@ -0,0 +1,108 @@
+  
+    - isolation - truly ensuring that one change cannot impact another. For example, our focus on API design and micro service architecture on the backend is one technique we use to create isolation
+    
+    - continuous delivery - by releasing small changes continuously to our production environment, we can minimze the impact of any one change and significantly increase time to resolution.
```
remove "can" => "we minimize"

mbryzek replied:
done

mikeclimbrock commented on Sdlc.md:
```diff
@@ -0,0 +1,108 @@
+    
+    - rollback - by ensuring that every application change can be rolled back, we enable faster resolution when problems arise. For example, our practice of managing database schema changes independently of application changes provides the means to ensure every application change can itself be rolled back.
+    
+    - incremental rollout - by slowly deploying application changes to production, and verifying that they work, we can minimize the impact of a problem before it affects all of our users
```
now sure "slowly" is correct word here.

chrhicks replied:
incrementally ?

mbryzek replied:
changed to by rolling out application changes to production incrementally

mikeclimbrock commented on Sdlc.md:
```diff
@@ -0,0 +1,108 @@
+
+Design of the system starts as soon as we begin discussing use cases and requirements. While the design stage starts early in our process, it is not concluded until the application is successfully completed - simple, stable, tested, in production, and meeting the needs of users.
+
+APIs designed for our customers further must meet the requirements of the 'api' repository - https://github.com/flowcommerce/api - and designed using [apidoc](apidoc.me/flow)
```
"further must meet" => "must additionally meet"

mbryzek replied:
chagned to must also meet

mikeclimbrock commented on Sdlc.md:
```diff
@@ -0,0 +1,108 @@
+
+1. Local development environment
+2. Local integration environment, supported by [workstation](https://github.com/flowcommerce/workstation)
+3. Production, supported by 
```
missing link to delta? [delta](https://github.com/flowcommerce/delta)

mbryzek replied:
done

mikeclimbrock commented on Sdlc.md:
```diff
@@ -0,0 +1,108 @@
+2. Local integration environment, supported by [workstation](https://github.com/flowcommerce/workstation)
+3. Production, supported by 
+
+In all cases, docker artifacts must be produced as these are the artifacts used in both the integration and production environments.
```
"are the artifacts" "artifacts are"

mbryzek replied:
done

mikeclimbrock commented on Sdlc.md:
```diff
@@ -0,0 +1,108 @@
+  - [Source code management](https://github.com/flowvault)
+  - [Contributions](https://github.com/flowvault/standards/Contributions.md)
+  
+We also have decided as a group that we prefer our applications to be developed with similar style - thus making it easy for a developer to move from one project to another. Thus you will see very similar style of development across all Play/Scala applications at Flow as you will see across all NodeJs applications.
```
"We also have decided as a group" => "As a group, we decided"

mikeclimbrock replied:
"similar style" => "similar styles"

mikeclimbrock replied:
"as you will see across" => "and across"

mbryzek replied:
done

mikeclimbrock commented on Sdlc.md:
```diff
@@ -0,0 +1,108 @@
+
+### Dependendency Management
+
+We strive to minimize the number of external dependencies. When we do bring in dependencies, we are resposible for reviewing the code as if it is our own - and we remain accountable for its performance and security at Flow.
```
"When we do bring in dependencies" => "When new dependencies are introduced"

mbryzek replied:
changed to When we introduce dependencies, we are resposible for reviewing the code as if it is our own and we remain accountable for its performance and security at Flow

mikeclimbrock commented on Sdlc.md:
```diff
@@ -0,0 +1,108 @@
+
+To consume third party APIs - our preference, where possible, is to first document the API using [apidoc](apidoc.me/flow) which then ensures we can consume the third party API without introducing new dependencies.
+
+To track third party dependencies, for Scala based applications, we use [dependency](dependency.flow.io) and for Node applications we use grasshopper. We commit to "OCD Fridays" where we spend the morning every week upgrading to the latest versions of all libraries. This practice of continuous upgrade aligns with our philosophy of continuous delivery.
```
add "," after "Node applications,"

chrhicks replied:
grasshopper - we use `npm` , specifically `npm outdated` now

mbryzek replied:
done

mikeclimbrock commented on Sdlc.md:
```diff
@@ -0,0 +1,108 @@
+
+All infrastructure runs on AWS, preferring the use of their shared services for storage (e.g. RDS).
+
+### Erors
```
"Errors"

mbryzek replied:
done

mikeclimbrock commented on Sdlc.md:
```diff
@@ -0,0 +1,108 @@
+
+When errors occur, teams should immediately respond:
+
+  - We always move forward; if necessary rollback by creating a new git tag
```
add "," after "if necessary,"

mbryzek replied:
done

mikeclimbrock commented on Sdlc.md:
```diff
@@ -0,0 +1,108 @@
+When errors occur, teams should immediately respond:
+
+  - We always move forward; if necessary rollback by creating a new git tag
+  - For any non trivial error, announce immediately in slack channel and make sure the people that can help are helping (e.g. interrupt any discussion or meeting, call them, etc)
```
"slack" => "Slack"

mbryzek replied:
done

mikeclimbrock commented:
- should there be any mention of adding an application to Registry?
- should there be any mention to configure integration builds in TravisCI (.com vs. .org)
- anything on publishing new artifacts to Artifactory?
- mention weekly standup?

chrhicks commented on Sdlc.md:
```diff
@@ -0,0 +1,108 @@
+
+We strive to minimize the number of external dependencies. When we do bring in dependencies, we are resposible for reviewing the code as if it is our own - and we remain accountable for its performance and security at Flow.
+
+To consume third party APIs - our preference, where possible, is to first document the API using [apidoc](apidoc.me/flow) which then ensures we can consume the third party API without introducing new dependencies.
```
remove `is to first document` -> `is to document`

mbryzek replied:
done

mbryzek commented:
> should there be any mention of adding an application to Registry?
Added

> should there be any mention to configure integration builds in TravisCI (.com vs. .org)
Added

> anything on publishing new artifacts to Artifactory?
Added

> mention weekly standup?
Added

mikeclimbrock commented on Sdlc.md:
```diff
@@ -0,0 +1,131 @@
+  - Unique port allocations, simplifying configuration for local integration environments
+  - Declared dependency which informs which other applications, including databases, need to be running and healthy for a given application to work
+
+Docker artifacts must be produced as these artifacts are used in both the integration and production environments. Docker artifacts are built of the master branch of each repository. Docker artifacts are stored and managed by [Docker Hub](http://hub.docker.com).
```
typo: "of"=>"off"

mikeclimbrock commented:
one more small comment, otherwise, LGTM!

mbryzek commented:
fixed typo
